## Quiz Questions 12 

In [13]:
import numpy as np
from pandas import DataFrame
from scipy import stats

# 1. Create the data
np.random.seed(12)

n = 100

wcc = np.round(np.random.normal(15, 5, n), 0)
crp = (wcc * 2) + np.round(np.random.normal(0, 10, n), 0)
lipase = wcc + crp + np.round(np.random.normal(2, 10, n), 0)

df = DataFrame({
    'WCC': wcc,
    'CRP': crp,
    'Lipase': lipase
})

# 2. Create model using only numpy
X = np.column_stack((np.ones(n), wcc, crp))
y = lipase

# 3. Calculate coefficients
beta = np.linalg.inv(X.T @ X) @ X.T @ y
y_pred = X @ beta
residuals = y - y_pred

# 4. Calculate R-squared
ss_res = np.sum(residuals ** 2)
ss_total = np.sum((y - np.mean(y)) ** 2)
r_squared = 1 - (ss_res / ss_total)

# 5. Calculate F statistic and p-value
p_best = 3
p_mean = 1

F = ((ss_total - ss_res) / (p_best - p_mean)) / (ss_res / (n - p_best))
f_p_value = 1 - stats.f.cdf(F, p_best - p_mean, n - p_best)

# 6. Calculate coefficient p-values
mse = ss_res / (n - p_best)
cov_beta = mse * np.linalg.inv(X.T @ X)
se_beta = np.sqrt(np.diag(cov_beta))
t_values = beta / se_beta
p_values = 2 * (1 - stats.t.cdf(np.abs(t_values), df=n-p_best))

# 7. Print answers
print("1. Model: Lipase ~ WCC + CRP")

print("2. Coefficients")
print("Intercept:", beta[0])
print("WCC:", beta[1])
print("CRP:", beta[2])

print("3. F statistic:", F)
print("4. F p-value:", f_p_value)

print("5. P-values")
print("Intercept:", p_values[0])
print("WCC:", p_values[1])
print("CRP:", p_values[2])

print("6. R-squared:", r_squared)

1. Model: Lipase ~ WCC + CRP
2. Coefficients
Intercept: 2.374090396539347
WCC: 0.542610209569723
CRP: 1.1826836477074814
3. F statistic: 157.28700016186937
4. F p-value: 0.0
5. P-values
Intercept: 0.4326286564782651
WCC: 0.046000988498571926
CRP: 0.0
6. R-squared: 0.7643194178356721


### Question 1 

The model predicts Lipase using WCC and CRP.

### Question 2

The overall F statistic is about 157.29. The F p-value is very small, about 3.61e-31, so the whole model is statistically significant.

### Question 3

The WCC coefficient is about 0.54, with p-value about 0.046, so WCC is significant and the CRP coefficient is about 1.18, with p-value close to 0, so CRP is highly significant.

### Question 4

The R² value is about 0.764, meaning the model explains about 76.4% of the variation in Lipase.